### DISCUSSION TITLE
Proposal: option to exclude `storage` from backups

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#4436] [Q&A]
Following up on https://github.com/pocketbase/pocketbase/discussions/4422#discussioncomment-8580918

As I think about it in more detail, I think there is a good use case for wanting to exclude uploaded files (`storage`) from backups.

I know of several users on PocketHost who have uploaded >20GB of what amounts to static files that change infrequently but that belong behind the API security layer provided by PocketBase. They would benefit from being able to perform data-only backups on a regular schedule.

Another thought - aren't files named uniquely based on content hash? So worst case, there would be extra/orphaned files in the file system, but never missing files? Is that true?

Again my personal use case for excluding storage is so I can use fuse mounts and effectively provide s3-backed storage without needing to manage/inject/expose s3 API keys into the instance settings.

Perhaps a checkbox `include uploaded files` would solve the case for data-only backups?

**Comment by ganigeorgiev** [comment]
> Another thought - aren't files named uniquely based on content hash?

No. The uploaded files name is constructed from the normalized original name + some random part.

> So worst case, there would be extra/orphaned files in the file system, but never missing files? Is that true?

Unless I'm misunderstanding what do you mean, no - there is no guarantee that there will not be missing files. In case a record is deleted after a backup the files associated to that record will be also deleted and if we don't backup the storage on restore they will be missing.

> They would benefit from being able to perform data-only backups on a regular schedule.
> ... Perhaps a checkbox include uploaded files would solve the case for data-only backups?

We can easily add a checkbox or a flag but I'm not sure what would be the use case for data-only backups.

If users backup only the database without the uploaded files there is no guarantee that their application will work in case of a restore. And as mentioned in the previous discussion, the fact that we don't backup S3 files is not a feature but a limitation because I just wasn't able to find an efficient way to do it and since most S3 vendors comes with their own backup/history/versioning tools I considered this "good enough".

In general the PocketBase auto backups may not the best option for recurring backup strategy if you have a lot of files (for example I don't use it for daily backups in Presentator as mentioned in https://github.com/pocketbase/pocketbase/discussions/4254).
They are simply full snapshot/archive of the `pb_data` at the moment the backup was requested (even if we introduce at later stage incremental backups we still will need at least 1 full snapshot to make the backup actually usable).

I'll have to think a little more on this.

What is the expected workflow in PocketHost in case users want to backup/restore?

**Reply by benallfree** [reply]
Users frequently want to restore a data snapshot from local dev, or pull a snapshot to local dev. They can do that via the admin or FTP. 

I agree that data integrity could be affected. On the other hand, there is currently no practical way via the admin to pull a data snapshot of a system with a very large storage directory. 

Maybe having them go directly to FTP is good enough, we would just need to make sure the instance is not running when they overwrite data.db. If that’s the case, I would likely choose to block API calls to the backup system to prevent large storage directories from causing issues.
